helpers/helpers.go

```diff
@@ -153,6 +156,70 @@ func GetIncludedBundles(filename string) ([]string, error) {
 	return includes, nil
 }
 
+// UnpackFile unpacks a .tar or .tar.gz/.tgz file to a given directory
+// Should be roughly equivalent to "tar -x[z]f file -C dest"
+func UnpackFile(file string, dest string) error {
+	fr, err := os.Open(file)
+	defer fr.Close()
+	if err != nil {
+		return err
+	}
+
+	var tr *tar.Reader
+
+	// If it's a compressed tarball
+	if strings.HasSuffix(file, ".tar.gz") || strings.HasSuffix(file, ".tgz") {
+		gzr, err := gzip.NewReader(fr)
+		defer gzr.Close()
+		if err != nil {
+			return errors.Wrapf(err, "Error decompressing tarball: %s", file)
+		}
+		tr = tar.NewReader(gzr)
+	} else {
+		 tr = tar.NewReader(fr)
+	}
+
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			// end of archive
+			break
+		} else if err != nil {
+			return errors.Wrapf(err, "Error reading contents of tarball: %s", file)
+		}
+
+		// Skip GitHub generated "extended header" file
+		if strings.Compare(hdr.Name, "pax_global_header") == 0 {
+			continue
+		}
+
+		out := filepath.Join(dest, hdr.Name)
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+			if _, err := os.Stat(out); err != nil {
+				if err = os.Mkdir(out, 0777); err != nil {
+					return errors.Wrapf(err, "Error unpacking directory: %s", out)
+				}
+			}
+		case tar.TypeReg:
+			of, err := os.OpenFile(out, os.O_CREATE|os.O_RDWR|os.O_EXCL, os.FileMode(hdr.Mode))
+			defer of.Close()
```

Comment:
I don't recommend using defer in this case. Just close "of" directly in the lines below right after the Copy happens, and before checking for the Copy error.


Defer works by adding a the function call to a "cleanup list" in the scope of the function (not the block). This means that we are only closing the files after all of them were opened.

Reply:
Alternative is to pull up the code to a function (either in local scope or toplevel). In this case, since you can resolve it with a single Close, don't think it is worth.

Reply:
Ah, that's a really good point about them all being opened at once.

As mentioned above, though, I still don't understand how doing it that way (after two different possible return statements) doesn't leave the possibility of the file hanging open after the function returned.

Reply:
Let me see if I can explain.

    of, err = os.OpenFile(...)

will either

1. Set `of` to `nil` and set `err` to something of type `error` or
1. Set `of` to be a pointer to some kind of data structure and set `err` to `nil`

In the second case there is nothing to close, so you just need to check `err` and if it is not `nil` return it. No `defer` is needed.

In the first case you do need to close `of`. However all you are doing before you want to close it is to call `io.Copy()` so the suggestion is that you write

      of, err := os.OpenFile(.....)
      if err != nil { return err ; }
      _, err = io.Copy()
      cerr := of.Close()
      if err != nil { return errors.Wrapf(err, ...) }
      if cerr != nil { return errors.Wrapf(cerr, "Probably out of disk space"....) }

(which means you also handle any errors you get when closing the output file that are currently missed).

Reply:
Definitely really appreciate you and @cmarcelo's help, here and on IRC. Still new to Go, but usually a quick learner ;)

BTW, thanks for reminding me that I'm ignoring the err return on `.Close()`. I noticed on c3fb081eafa5bdfe31762e45312d6c00529d1c67 that @tmarcu replaced all `defer variable.Close()` with
```go
defer func() {
 	_ = variable.Close()
 }()
```
to get rid of linting warnings. Guessing he probably wants me to do the same.

While I'm at it, I'm going to go through the Travis output to see all the other places linting yells at me.

Reply:
@kevincwells You do know how to install the linter locally?

Handling errors associated with close is always something for debate.Traditionally the C stdio library fclose function causes a flush of the buffer, and that is a place where you can get 'out of disk' or 'out of quota' errors returned from write.  The actual errors that get returned from close(2) are very limited, essentially EIO, which there is almost nothing you can do about. With go, you are encouraged to do the flush yourself, to make the error handling more explicit.